url146c819f25_jlink_Parse4SJM_tgz-pJ8-sjm_examples_book_robot_ShowRobotRefactoredJ8/src/sjm/examples/book/query/ComparisonParser.java

```diff
@@ -0,0 +1,196 @@
+package sjm.examples.book.query;
+
+import sjm.examples.book.logic.*;
+import sjm.parse.*;
+import sjm.parse.tokens.*;
+
+/**
+ * This utility class provides support to the Jaql 
+ * parser, specifically for <code>expression()</code>
+ * and <code>comparison()</code> subparsers.
+ *
+ * The grammar this class supports is:
+ *
+ * <blockquote><pre>	
+ *     comparison = arg operator arg;
+ *     arg        = expression | QuotedString;
+ *     expression = term ('+' term | '-' term)*;
+ *     term       = factor ('*' factor | '/' factor)*;
+ *     factor     = '(' expression ')' | Num | variable;
+ *     variable   = Word;
+ *     operator   = "<" | ">" | "=" | "<=" | ">=" | "!=";
+ * </pre></blockquote>
+ * 
+ * @author Steven J. Metsker
+ *
+ * @version 1.0 
+ */
+public class ComparisonParser {
+	protected Sequence expression;
+	protected Speller speller;
+
+	/**
+	 * Construct a ComparisonParser that will consult the
+	 * given speller for the proper spelling of variable
+	 * names.
+	 */
+	public ComparisonParser(Speller speller) {
+		this.speller = speller;
+	}
+
+	/**
+	 * Returns a parser that will recognize a comparison
+	 * argument.
+	 */
+	public Parser arg() {
+
+		// arg = expression | QuotedString;
+
+		Alternation a = new Alternation();
+		a.add(expression());
+		a.add(new QuotedString().setAssembler(new AtomAssembler()));
+		return a;
+	}
+
+	/**
+	 * Returns a parser that will recognize a comparison.
+	 */
+	public Parser comparison() {
+		Sequence s = new Sequence("comparison");
+		s.add(arg());
+		s.add(operator());
+		s.add(arg());
+		s.setAssembler(new ComparisonAssembler());
+		return s;
+	}
+
+	/*
+	 * Recognize '/' followed by a factor.
+	 */
+	protected Parser divideFactor() {
+		Sequence s = new Sequence("divideFactor");
+		s.add(new Symbol('/').discard());
+		s.add(factor());
+		s.setAssembler(new ArithmeticAssembler('/'));
+		return s;
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 71:75, 128:132, 153:157, 183:187.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.